plugins/pay.c

```diff
@@ -372,24 +419,56 @@ static struct command_result *getroute_done(struct command *cmd,
 	feepercent = ((double)fee) * 100.0 / ((double) pc->msatoshi);
 
 	if (fee > pc->exemptfee && feepercent > pc->maxfeepercent) {
+		const jsmntok_t *charger;
+
 		attempt_failed_fmt(pc, "{ 'message': 'Route wanted fee of %"PRIu64" msatoshis' }", fee);
 
+		/* Remember this if we eliminating this causes us to have no
+		 * routes at all! */
+		if (!pc->expensive_route)
+			pc->expensive_route
+				= tal_fmt(pc, "Route wanted fee of %"PRIu64
+					  " msatoshis", fee);
+
+		/* Try excluding most fee-charging channel (unless it's in
+		 * routeboost). */
+		charger = find_worst_channel(buf, t, "msatoshi", pc->msatoshi);
+		if (maybe_exclude(pc, buf, charger))
+			return start_pay_attempt(cmd, pc);
+
 		if (tal_count(pc->routehints) != 0)
 			return next_routehint(cmd, pc);
 
 		return command_fail(cmd, PAY_ROUTE_TOO_EXPENSIVE,
-				    "Route wanted fee of %"PRIu64" msatoshis",
-				    fee);
+				    "%s", pc->expensive_route);
 	}
 
 	if (delay > pc->maxdelay) {
-		attempt_failed_fmt(pc, "{ 'message': 'Route wanted delay %u blocks' }", delay);
+		const jsmntok_t *delayer;
+
+		attempt_failed_fmt(pc,
+				   "{ 'message': 'Route wanted delay of %u blocks' }",
+				   delay);
+
+		/* Remember this if we eliminating this causes us to have no
+		 * routes at all! */
+		if (!pc->expensive_route)
+			pc->expensive_route
+				= tal_fmt(pc, "Route wanted delay of %u blocks",
+					  delay);
+
+		delayer = find_worst_channel(buf, t, "delay", pc->final_cltv);
```

Comment:
And I think the same is true here, we are eliminating not the most expensive hop, but the last one since delays accumulate.